cloud migration rc1

filipesilva opened the pull request:
Fix https://github.com/metabase/metabase/issues/43056

filipesilva committed:
fix: only set migration version on dev tag

filipesilva committed:
fix: fe/be should agree on using staging migration

filipesilva committed:
feat: show errors on migration creation

npfitz commented on frontend/src/metabase/admin/settings/components/CloudPanel/utils.ts:
```diff
@@ -71,9 +71,11 @@ export const getMigrationEventTime = (isoString: string) =>
-    : `https://store.staging.metabase.com`;
+  // Only used from components, but passing the setting value from all callers is a PITA.
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  const baseUrl = useSetting("migration-use-staging")
```
Instead of this, can we just have a setting that provides the store URL? That way we don't need to keep the FE and BE values in sync if they ever change

npfitz replied:
Yeah, either name would probably be fine.

filipesilva replied:
Both are needed, because the BE uses `https://store-api.metabase.com/` and the FE uses `https://store.metabase.com/`. I added a single flag so we could control in a single way, since they need to agree.

Consider testing the JAR: with the single flag I have right now you'd do `MB_MIGRATION_USE_STAGING=true java -jar metabase.jar`.

But with url settings you'd do something like `MB_MIGRATION_BE_URL=https://store-api.metabase.com MB_MIGRATION_FE_URL=https://store.metabase.com java -jar metabase.jar`.

Thinking through it, I think adding the separate flags is worse. Maybe a setting just for the domain (`metabase.com` vs `staging.metabase.com`) would be better?

filipesilva replied:
Thought about it a bit more and I think setting the domain isn't great because it wouldn't allow actually changing the domain in some cases. For instance when developing locally, because you still need to know it's http vs https. And I know I need this for an upcoming project already so I can test store integration without deploying first.

I think I'll add the URL settings, but make their default value depend on another setting like `MB_STORE_USE_STAGING` for ease of jar testing.

filipesilva replied:
Done

npfitz commented on frontend/src/metabase/admin/settings/components/CloudPanel/CloudPanel.tsx:
```diff
@@ -104,6 +105,10 @@ export const CloudPanel = ({
           )}
+
+          {createCloudMigrationResult.isError && (
+            <MigrationCreationError error={createCloudMigrationResult.error as MigrationCreationErrorProps} />
```
We generally try to avoid typecasting when possible. I can take a look and see if I can resolve the type issues if you'd like

filipesilva replied:
My editor was saying this was a `unknown`, which really surprised me. But yeah feel free to, or I will later.

filipesilva replied:
Saw another component also using a [result error](https://github.com/metabase/metabase/blob/057cb3f38677fa6bfa14408f7c6919422d8856c1/frontend/src/metabase/admin/performance/components/StrategyEditorForDatabases.tsx#L125). Used the same casting (`any`) on the props.

filipesilva committed:
fix: h2 dumps should not be read-only

It breaks key rotation during dump and on rotate-encryption-key calls.

filipesilva committed:
fix: don't ignore errors during key rotation

rotate-encryption-key! has nice errors but we ignore them.

filipesilva committed:
fix: use tmp file for cloud migration dump

The docker process runs as non-root and doesn't have access to all folders.
But it does have access to tmp folders, and other places like metabase.email.messages already make use of it.

Plus, it's semantically more correct.

filipesilva commented on frontend/src/metabase/admin/settings/components/CloudPanel/MigrationCreationError.tsx:
```diff
@@ -0,0 +1,36 @@
+import { LargeIconContainer, MigrationCard } from "./CloudPanel.styled";
+
+interface MigrationCreationErrorProps {
+  error: {data?: string, status: number};
```
Removing, can always add back later.

filipesilva committed:
feat: store-url and store-api-url settings

filipesilva committed:
fix: get-store-migration should be using migration-url

filipesilva commented on frontend/src/metabase-types/api/settings.ts:
```diff
@@ -165,7 +165,7 @@ export const tokenFeatures = [
 ] as const;
 
-export type TokenFeature = typeof tokenFeatures[number];
+export type TokenFeature = (typeof tokenFeatures)[number];
```
prettier did this

filipesilva commented on frontend/src/metabase/admin/settings/components/CloudPanel/utils.ts:
```diff
@@ -71,9 +71,9 @@ export const getMigrationEventTime = (isoString: string) =>
-    : `https://store.staging.metabase.com`;
+  // Only used from components, but passing the setting value from all callers is a PITA.
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  const baseUrl = useSetting("store-url");
```
Failing atm, because of the hooks stuff. Nick taking a look.

filipesilva commented on src/metabase/cmd.clj:
```diff
@@ -239,8 +239,8 @@
-    (catch Throwable _e
-      (log/error "ERROR ROTATING KEY.")
+    (catch Throwable e
+      (log/error e "ERROR ROTATING KEY.")
```
Discussed on slack https://metaboat.slack.com/archives/CKZEMT1MJ/p1716575795409719?thread_ts=1716561641.036369&cid=CKZEMT1MJ

filipesilva commented on src/metabase/cmd/copy.clj:
```diff
@@ -164,6 +164,10 @@
+    :model/Setting
+    ;; Never create dumps with read-only-mode turned on.
+    ;; It will be confusing to restore from and prevent key rotation.
+    (remove (fn [{k :key}] (= k "read-only-mode")))
```
@dpsutton 's solution to read-only dumps.

escherize replied:
I don't understand this comment. It seems like this remove xform on 170 will filter out the "read-only-mode" setting, but it will still be possible to create dumps with read-only-mode turned on, right? (it'll just not dump that setting).

I might be missing something, though

escherize replied:
Maybe I can parse the comment now, so this will not serialize the read-only-mode setting, and it will end up in the default (false) state. Would you consider just making it `false` in the xformer instead of removing it? That would be a little more direct.

filipesilva replied:
I'm not super confident that's strictly better, because it means committing some data to a written format and it also shouldn't be a factor for "normal" dumps. Since the end result is the same, I'd rather default on the conservative side and not add new data to dumps.

filipesilva commented on src/metabase/models/cloud_migration.clj:
```diff
@@ -33,29 +33,42 @@
+
+(defsetting store-url
+  (deferred-tru "Store URL.")
+  :visibility :admin ;; should be :internal, but FE doesn't get internal settings
```
This setting should be `:internal` but the FE doesn't get internal settings. I could make it get them, but I'm not confident that changing visibility logic is the better approach here. Defaulted to the low effort approach.

@dpsutton WDYT?

escherize replied:
It seems like `:admin` is what you want, given the FE needs read the setting, but why do you think it should be `:internal`?

filipesilva replied:
I'm trying to tell the FE what the store url is, in a way that I can alter it both on local development environments and when testing jar/docker artifacts.

noahmoss replied:
I think `:admin` is expected here if it needs to be surfaced to the FE. It means that it'll only be surfaced to admin users and not non-admins.

If you want to ensure it's not _set_ by admins, set the `:setter` to `:none`.

filipesilva committed:
fix: remove change introduced by prettier that breaks prettier check

I think? CI just tells me something is wrong.
But running `yarn prettier` seems to be changing a hundred files, so it looks off.

filipesilva committed:
fix: missing setting mock

filipesilva committed:
fix: join dump file dir using io/file

On mac it ends with /, but unix it doesn't. Safer this way.

npfitz committed:
store-url setting

filipesilva committed:
test: use prop mocks in cloud panel test

filipesilva committed:
test: fix progress flake

Fix #43207

filipesilva committed:
test: add onMigrationStart back for tests

escherize commented on src/metabase/models/cloud_migration.clj:
```diff
@@ -229,7 +240,8 @@
   ;; dump-to-h2 starts behaving oddly if you try to dump repeatly to the same file
   ;; in the same process.
-  (let [dump-file (io/file (str "cloud_migration_dump_" (random-uuid) ".mv.db"))]
+  (let [dump-file (io/file (System/getProperty "java.io.tmpdir")
+                           (str "cloud_migration_dump_" (random-uuid) ".mv.db"))]
```
How come we're dumping to a temp directory now? Maybe update the comment if there's a good reason

filipesilva replied:
Added:
```
  ;; The docker image process runs in non-root, so write to a dir it can access.
```

Was a rather nasty error mode actually.

escherize commented on src/metabase/models/cloud_migration.clj:
```diff
@@ -98,17 +106,20 @@
-  (str (migration-store-url) "/" external-id path))
+(defn migration-url
+  "Store API URL for migrations."
+  ([]
+   (str (store-api-url) "/api/v2/migration"))
```
nit: This arity is only used once, what do you think about inlining the 1-arity version of `migration-url`?

filipesilva replied:
It's used twice: once in itself (`migration-url`) to build the base part of the url and then add the path, and once in `get-store-migration`. Inlining would be a bit of a semantic loss and more verbose here I think.

escherize replied:
Makes sense, thanks. Also, really glad to see the versioned endpoint. 🙏

escherize reviewed (changes requested):
Overall the BE portion looks pretty good, I had a couple questions, partially to help me understand the "why's". 


I would like the question on https://github.com/metabase/metabase/pull/43082/files#diff-65fdb622c6760fde1867daf5516614006d0c0fb106cf63f5718cd31bde2ebbefR170 explained, though. I do think setting it to be false makes more sense.

I'll stamp it once that's explained.

filipesilva commented on src/metabase/models/cloud_migration.clj:
```diff
@@ -33,29 +33,42 @@
   ;; This will cause the restore to fail on cloud unless you also set `migration-dump-file` to
   ;; a dump from that version, but it lets you test everything else up to that point works.
-  :default    (when config/is-dev? "v0.49.7")
+  :default    (when (= (config/mb-version-info :tag) "vLOCAL_DEV") "v0.50.0-RC1")
```
To migrate to cloud you need to use a version that cloud knows about, so that the cloud instance is started with at least that version. Cloud doesn't run 50 yet because it's not released. RC1 is there though, so we can use that.

I haven't fully worked out how this should work for the general purpose dev. I mean, it can't always work - if you have migrations on your branch and there is no release yet, using an older version on cloud will fail when restoring from the dump.

But setting a known version works up to that point. I think there's more sophisticated things that can be done here, like [auto-deploying PR builds to cloud](https://www.notion.so/metabase/How-to-create-an-internal-cloud-instance-dc27b9d8dd494fc8a3abd08b9ff1e13d?pvs=4#d9258689b74748a69a4aea3ba682ab75) and using the branch name as the version. But that's a separate piece of work.

filipesilva committed:
doc: add comment about tmp dir for dumps

escherize reviewed (approved):
Looks good. Thanks for your explanations
